openssl/src/crypto/signature.rs

```diff
@@ -0,0 +1,289 @@
+//! be used to verify the integrity and authenticity of signed messages
+//! using the `Verifier`.
+//!
+//! # Example
+//!
+//! ```rust
+//! use std::io::prelude::*;
+//! use openssl::crypto::signature::{Signer, Verifier};
+//! use openssl::crypto::pkey::PKey;
+//! use openssl::crypto::hash;
+//!
+//! // Generate public/private keypairs
+//! let mut keypair = PKey::new();
+//! let mut privkey = PKey::new();
+//! let mut pubkey = PKey::new();
+//! keypair.gen(1024);
+//! privkey.load_priv(&keypair.save_priv()[..]);
+//! pubkey.load_pub(&keypair.save_pub()[..]);
+//!
+//! // generate some data
+//! let data1: Vec<u8> = (0..25).cycle().take(1024).collect();
+//! let data2: Vec<u8> = (100..150).cycle().take(1024 * 5).collect();
+//!
+//! // sign some content with private key
+//! let mut signer = Signer::new(hash::Type::SHA512, &privkey).unwrap();
+//! signer.write_all(&data1[..]).unwrap();
+//! signer.write_all(&data2[..]).unwrap();
+//! let signature = signer.finish().unwrap();
+//!
+//! // verify content using signature and public key
+//! let mut verifier = Verifier::new(hash::Type::SHA512, &pubkey).unwrap();
+//! verifier.write_all(&data1[..]).unwrap();
+//! verifier.write_all(&data2[..]).unwrap();
+//! assert!(verifier.finish(&signature[..]).unwrap());
+//! ```
+
+use std::ptr;
+use std::io::{self, Write};
+use std::marker::PhantomData;
+use libc;
+
+use ffi;
+use ssl::error::SslError;
+use crypto::pkey::PKey;
+use crypto::hash;
+
+/// Signature Generation Context
+pub struct Signer<'a> {
+    ctx: *mut ffi::EVP_MD_CTX,
+    pkey: PhantomData<&'a PKey>,
+}
+
+/// Signature Verification Context
+pub struct Verifier<'a> {
+    ctx: *mut ffi::EVP_MD_CTX,
+    pkey: PhantomData<&'a PKey>,
+}
+
+impl<'a> Signer<'a> {
+    /// Create and initialize a new Signer
+    ///
+    /// The digest type specified will determine the final form  of the
+    /// signature returned by `finalize`.
+    pub fn new(digest_type: hash::Type, pkey: &'a PKey) -> Result<Signer, SslError> {
+        ffi::init();
+
+        let ctx = unsafe {
+            let r = ffi::EVP_MD_CTX_create();
+            assert!(!r.is_null(), "EVP_MD_CTX_create failed");
+            r
+        };
+
+        let md = digest_type.evp_md();
+
+        unsafe {
+            try_ssl_if!(1 !=
+                        ffi::EVP_DigestSignInit(ctx,
+                                                ptr::null_mut(),
+                                                md,
+                                                ptr::null(),
+                                                pkey.get_handle()));
+        }
+
+        Ok(Signer {
+            ctx: ctx,
+            pkey: PhantomData,
+        })
+    }
+
+    /// Feed bytes to be added to the signature calculation
+    #[inline]
+    fn update(&mut self, data: &[u8]) -> Result<(), SslError> {
+        unsafe {
+            lift_ssl_if!(1 != ffi::EVP_DigestUpdate(self.ctx, data.as_ptr(), data.len() as u32))
```

Comment:
> This should loop over data in case it's longer than u32 bytes. Check out the hash module for a reference.

Agreed, if I make this public.  Previously, this was dealt with at the public interface layer (`Write` impl).